espnet/nets/pytorch_backend/transformer/attention.py

```diff
@@ -142,24 +144,32 @@ def rel_shift(self, x, zero_triu=False):
         """Compute relative positinal encoding.
 
         Args:
-            x (torch.Tensor): Input tensor (batch, time, size).
+            x (torch.Tensor): Input tensor (batch, head, time1, time2).
             zero_triu (bool): If true, return the lower triangular part of the matrix.
 
         Returns:
             torch.Tensor: Output tensor.
 
         """
-        zero_pad = torch.zeros((*x.size()[:3], 1), device=x.device, dtype=x.dtype)
-        x_padded = torch.cat([zero_pad, x], dim=-1)
-
-        x_padded = x_padded.view(*x.size()[:2], x.size(3) + 1, x.size(2))
-        x = x_padded[:, :, 1:].view_as(x)
+        idx = torch.arange(0, x.size(3), device=x.device)
+        k_idx, q_idx = idx.unsqueeze(0), idx.unsqueeze(1)
+        # for pos_idx[i, j], i means the postion of query, j means the position of key
+        # the range of i-j is from -(L-1) to L-1
+        # use the absolute value of i-j, so the range is from 0 to L-1
+        rel_pos_idx = torch.abs(k_idx - q_idx)
+        # original postional encodings are generated with reversed order
+        rel_pos_idx = x.size(3) - 1 - rel_pos_idx
+        rel_pos_idx = rel_pos_idx.repeat(x.size(0), x.size(1), 1, 1)
+        x_shift = torch.gather(x, dim=3, index=rel_pos_idx)
```

Comment:
```suggestion
        x_shift = torch.gather(x, dim=2, index=rel_pos_idx)
```
Let's think  a case of `qlen != klen`.